// dojo-compose/mixins/createStateful.js
(function (factory) {
    if (typeof module === 'object' && typeof module.exports === 'object') {
        var v = factory(require, exports); if (v !== undefined) module.exports = v;
    }
    else if (typeof define === 'function' && define.amd) {
        define(["require", "exports", 'dojo-core/lang', 'dojo-core/WeakMap', './createEvented', '../compose'], factory);
    }
})(function (require, exports) {
    "use strict";
    var lang_1 = require('dojo-core/lang');
    var WeakMap_1 = require('dojo-core/WeakMap');
    var createEvented_1 = require('./createEvented');
    var compose_1 = require('../compose');
    /**
     * A weak map of stateful instances to their obseved state references
     */
    var observedStateMap = new WeakMap_1.default();
    /**
     * Internal function to unobserve the state of a `Stateful`
     * @param stateful The `Stateful` object to unobserve
     */
    function unobserve(stateful) {
        var observedState = observedStateMap.get(stateful);
        if (observedState) {
            observedState.handle.destroy();
            stateful.emit({
                type: 'observe:complete',
                target: stateful
            });
        }
    }
    /**
     * Internal function that actually applies the state to the Stateful's state and
     * emits the `statechange` event.
     * @param stateful The Stateful instance
     * @param state The State to be set
     */
    function setStatefulState(stateful, state) {
        state = lang_1.deepAssign(stateWeakMap.get(stateful), state);
        stateful.emit({
            type: 'statechange',
            state: state,
            target: stateful
        });
    }
    /**
     * A weak map that contains the stateful's state
     */
    var stateWeakMap = new WeakMap_1.default();
    /**
     * Create an instance of a stateful object
     */
    var createStateful = compose_1.default({
        get state() {
            return stateWeakMap.get(this);
        },
        setState: function (value) {
            var stateful = this;
            var observedState = observedStateMap.get(stateful);
            if (observedState) {
                observedState.observable.patch(value, { id: observedState.id });
            }
            else {
                setStatefulState(stateful, value);
            }
        },
        observeState: function (id, observable) {
            var stateful = this;
            var observedState = observedStateMap.get(stateful);
            if (observedState) {
                if (observedState.id === id && observedState.observable === observable) {
                    return observedState.handle;
                }
                throw new Error("Already observing state with ID '" + observedState.id + "'");
            }
            observedState = {
                id: id,
                observable: observable,
                subscription: observable
                    .observe(id)
                    .subscribe(function (item) { return setStatefulState(stateful, item); }, /* next handler */ function (err) {
                    /* TODO: Should we emit an error, instead of throwing? */
                    throw err;
                }, /* error handler */ function () { return unobserve(stateful); }),
                handle: {
                    destroy: function () {
                        var observedState = observedStateMap.get(stateful);
                        if (observedState) {
                            observedState.subscription.unsubscribe();
                            observedStateMap.delete(stateful);
                        }
                    }
                }
            };
            observedStateMap.set(stateful, observedState);
            return observedState.handle;
        }
    })
        .mixin({
        mixin: createEvented_1.default,
        initialize: function (instance, options) {
            var state = {};
            stateWeakMap.set(instance, state);
            instance.own({
                destroy: function () {
                    stateWeakMap.delete(instance);
                }
            });
            if (options) {
                if (options.state) {
                    instance.setState(options.state);
                }
                if (options.id && options.stateFrom) {
                    instance.own(instance.observeState(options.id, options.stateFrom));
                }
                else if (options.id || options.stateFrom) {
                    throw new TypeError('Factory requires options "id" and "stateFrom" to be supplied together.');
                }
            }
        }
    });
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.default = createStateful;
});
