// dojo-widgets/mixins/createListMixin.js
(function (factory) {
    if (typeof module === 'object' && typeof module.exports === 'object') {
        var v = factory(require, exports); if (v !== undefined) module.exports = v;
    }
    else if (typeof define === 'function' && define.amd) {
        define(["require", "exports", 'maquette/maquette', './createCachedRenderMixin'], factory);
    }
})(function (require, exports) {
    "use strict";
    var maquette_1 = require('maquette/maquette');
    var createCachedRenderMixin_1 = require('./createCachedRenderMixin');
    var createListMixin = createCachedRenderMixin_1.default
        .mixin({
        mixin: {
            getChildrenNodes: function () {
                var list = this;
                if (list.state && list.state.items) {
                    var items = list.state.items;
                    return [maquette_1.h(list.tagNames.list, items.map(function (item) { return maquette_1.h(list.tagNames.item, { key: item }, [item.label]); }))];
                }
                return [];
            },
            tagName: 'dojo-list',
            tagNames: {
                list: 'ul',
                item: 'li'
            }
        }
    });
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.default = createListMixin;
});
